File: test_parse_response.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
测试解析 data/response_body.html
包含完整的去重、数据库保存功能
"""

import os
import json
import re
import html
from datetime import datetime
from pathlib import Path
import base64

# 配置
class Config:
    BASE_DIR = Path(__file__).parent
    DATA_DIR = BASE_DIR / 'data'
    DATABASE_FILE = DATA_DIR / 'play_history_db.json'

# 确保数据目录存在
def ensure_data_dir():
    Config.DATA_DIR.mkdir(parents=True, exist_ok=True)

# 生成记录唯一ID
def generate_record_id(record):
    if record.get('trackId'):
        return f"track_{record['trackId']}"

    # 备用方案：使用曲目+难度+日期+达成率组合
    key = f"{record['title']}_{record['diff']}_{record['date']}_{record['percentage']}"
    # 使用base64编码并移除特殊字符
    encoded = base64.b64encode(key.encode('utf-8')).decode('utf-8')
    encoded = encoded.replace('=', '').replace('+', '').replace('/', '')
    return f"combo_{encoded}"

# 加载数据库
def load_database():
    ensure_data_dir()

    try:
        if Config.DATABASE_FILE.exists():
            with open(Config.DATABASE_FILE, 'r', encoding='utf-8') as f:
                data = json.load(f)
            print(f"✅ 数据库加载成功，已有 {len(data['records'])} 条记录\n")
            return data
    except Exception as e:
        print(f"⚠️  数据库加载失败: {e}，将创建新数据库\n")

    # 返回空数据库结构
    return {
        'version': 1,
        'createdAt': datetime.now().isoformat(),
        'lastUpdated': None,
        'totalFetches': 0,
        'records': [],
        'recordIndex': {}  # 用于快速查找的索引
    }

# 保存数据库
def save_database(db):
    ensure_data_dir()
    db['lastUpdated'] = datetime.now().isoformat()

    with open(Config.DATABASE_FILE, 'w', encoding='utf-8') as f:
        json.dump(db, f, ensure_ascii=False, indent=2)

    print(f"✅ 数据库已保存，共 {len(db['records'])} 条记录")

# 将新记录添加到数据库（去重）
def add_records_to_database(db, new_records):
    added_records = []
    duplicate_count = 0

    for record in new_records:
        record_id = generate_record_id(record)

        # 检查是否已存在
        if record_id in db['recordIndex']:
            duplicate_count += 1
            continue

        # 添加元数据
        enriched_record = {
            **record,
            '_id': record_id,
            '_fetchedAt': datetime.now().isoformat()
        }

        # 添加到数据库
        db['records'].append(enriched_record)
        db['recordIndex'][record_id] = True
        added_records.append(enriched_record)

    db['totalFetches'] += 1

    return {
        'addedRecords': added_records,
        'duplicateCount': duplicate_count
    }

# 解析HTML
def parse_records_html(html_content):
    print(f"HTML长度: {len(html_content)} 字符\n")

    records = []

    # 匹配每个游玩记录块
    record_block_pattern = r'<div class="p_10 t_l f_0 v_b">([\s\S]*?)(?=<div class="p_10 t_l f_0 v_b">|<div class="f_0">[\s\S]*?</footer>)'

    for match in re.finditer(record_block_pattern, html_content):
        block = match.group(1)

        try:
            # 提取难度
            diff_match = re.search(r'diff_(\w+)\.png', block)
            diff_map = {
                'basic': 'Basic',
                'advanced': 'Advanced',
                'expert': 'Expert',
                'master': 'Master',
                'remaster': 'Re:Master'
            }
            diff = diff_map.get(diff_match.group(1), diff_match.group(1)) if diff_match else 'Unknown'

            # 提取日期时间
            date_match = re.search(r'<span class="v_b">(\d{4}/\d{2}/\d{2} \d{2}:\d{2})</span>', block)
            date = date_match.group(1) if date_match else ''

            # 提取曲目名称
            title_match = re.search(r'clear\.png"[^>]*/>([\s\S]*?)</div>', block)
            title = ''
            if title_match:
                title = re.sub(r'<[^>]+>', '', title_match.group(1)).strip()
                # 解码HTML实体
                title = html.unescape(title)

            # 提取达成率
            achieve_match = re.search(r'playlog_achievement_txt[^>]*>(\d+)<span[^>]*>\.([\d]+)%</span>', block)
            percentage = f"{achieve_match.group(1)}.{achieve_match.group(2)}" if achieve_match else ''

            # 提取评级
            rate_match = re.search(r'playlog/(\w+)\.png[^>]*class="playlog_scorerank"', block)
            rate = ''
            if rate_match:
                rate_map = {
                    'sssp': 'SSS+', 'sss': 'SSS', 'ssp': 'SS+', 'ss': 'SS',
                    'sp': 'S+', 's': 'S', 'aaa': 'AAA', 'aa': 'AA', 'a': 'A',
                    'bbb': 'BBB', 'bb': 'BB', 'b': 'B', 'c': 'C', 'd': 'D'
                }
                rate = rate_map.get(rate_match.group(1).lower(), rate_match.group(1).upper())

            # 提取DX分数
            dx_match = re.search(r'white p_r_5 f_15 f_r">([\d,]+)\s*/\s*([\d,]+)</div>', block)
            dxscore = dx_match.group(1).replace(',', '') if dx_match else ''
            dxmax = dx_match.group(2).replace(',', '') if dx_match else ''

            # 提取FC状态
            fc = None
            fc_match = re.search(r'playlog/(fc_|ap)(\w*)\.png', block)
            if fc_match and 'dummy' not in fc_match.group(0):
                fc_type = fc_match.group(1) + fc_match.group(2)
                fc_map = {'ap': 'AP', 'app': 'AP+', 'fc': 'FC', 'fcp': 'FC+'}
                fc = fc_map.get(fc_type.lower(), fc_type.upper())

            # 提取FS状态
            fs = None
            fs_match = re.search(r'playlog/(sync_|fs|fsd)(\w*)\.png', block)
            if fs_match and 'dummy' not in fs_match.group(0):
                fs_type = fs_match.group(1) + fs_match.group(2)
                fs_map = {'sync': 'SYNC', 'fs': 'FS', 'fsp': 'FS+', 'fsd': 'FDX', 'fsdp': 'FDX+'}
                fs = fs_map.get(fs_type.lower(), fs_type.upper())

            # 提取记录ID
            id_match = re.search(r'name="idx" value="([^"]+)"', block)
            track_id = id_match.group(1) if id_match else ''

            # 是否是DX谱面
            is_dx = 'music_dx.png' in block

            # 是否New Record
            is_new_record = 'newrecord.png' in block

            if title:
                records.append({
                    'title': title,
                    'diff': diff,
                    'percentage': percentage,
                    'rate': rate,
                    'date': date,
                    'fc': fc,
                    'fs': fs,
                    'dxscore': dxscore,
                    'dxmax': dxmax,
                    'trackId': track_id,
                    'isDx': is_dx,
                    'isNewRecord': is_new_record
                })

        except Exception as e:
            print(f"⚠️  解析单条记录失败: {e}")

    return records

# 主函数
def main():
    print('=' * 60)
    print('测试解析 data/response_body.html（带去重功能）')
    print('=' * 60)
    print()

    html_file = Config.DATA_DIR / 'response_body.html'

    if not html_file.exists():
        print(f"❌ 文件不存在: {html_file}")
        return

    # 步骤1：加载数据库
    print('步骤1: 加载历史数据库')
    db = load_database()

    # 步骤2：读取和解析HTML
    print('步骤2: 读取和解析HTML')
    with open(html_file, 'r', encoding='utf-8') as f:
        html_content = f.read()

    records = parse_records_html(html_content)

    if not records:
        print('⚠️  未找到任何记录')
        return

    print(f"✅ 成功解析 {len(records)} 条游玩记录\n")

    # 步骤3：去重并添加到数据库
    print('步骤3: 去重并更新数据库')
    result = add_records_to_database(db, records)
    added_records = result['addedRecords']
    duplicate_count = result['duplicateCount']

    print(f"本次解析: {len(records)} 条")
    print(f"新增记录: {len(added_records)} 条")
    print(f"重复记录: {duplicate_count} 条（已跳过）")
    print(f"数据库总计: {len(db['records'])} 条\n")

    # 显示新增记录
    if added_records:
        print('=' * 60)
        print('新增记录（前10条）')
        print('=' * 60)
        print()

        display_count = min(10, len(added_records))
        for i in range(display_count):
            r = added_records[i]
            print(f"{i + 1}. {r['title']}")
            print(f"   难度: {r['diff']} | 日期: {r['date']}")
            print(f"   达成率: {r['percentage']}% | 评级: {r['rate']}")
            print(f"   DX分数: {r['dxscore']}/{r['dxmax']}")
            print()

        if len(added_records) > display_count:
            print(f"... 还有 {len(added_records) - display_count} 条新记录\n")
    else:
        print('ℹ️  所有记录都已存在，没有新增记录\n')

    # 统计信息
    print('=' * 60)
    print('统计信息（数据库总计）')
    print('=' * 60)

    stats = {
        'total': len(db['records']),
        'difficulties': {},
        'ratings': {},
        'fcCount': 0,
        'fsCount': 0,
        'newRecords': 0,
        'dxCount': 0
    }

    for r in db['records']:
        # 难度分布
        diff = r['diff']
        stats['difficulties'][diff] = stats['difficulties'].get(diff, 0) + 1

        # 评级分布
        if r['rate']:
            rate = r['rate']
            stats['ratings'][rate] = stats['ratings'].get(rate, 0) + 1

        # 计数
        if r['fc']:
            stats['fcCount'] += 1
        if r['fs']:
            stats['fsCount'] += 1
        if r['isNewRecord']:
            stats['newRecords'] += 1
        if r['isDx']:
            stats['dxCount'] += 1

    print(f"总记录数: {stats['total']}")
    print(f"DX谱面: {stats['dxCount']}")
    print(f"新纪录: {stats['newRecords']}")
    print(f"FC数量: {stats['fcCount']}")
    print(f"FS数量: {stats['fsCount']}")

    print('\n难度分布:')
    diff_order = ['Basic', 'Advanced', 'Expert', 'Master', 'Re:Master']
    for diff in diff_order:
        if diff in stats['difficulties']:
            print(f"  {diff}: {stats['difficulties'][diff]}")

    print('\n评级分布:')
    rating_order = ['SSS+', 'SSS', 'SS+', 'SS', 'S+', 'S', 'AAA', 'AA', 'A']
    for rate in rating_order:
        if rate in stats['ratings']:
            print(f"  {rate}: {stats['ratings'][rate]}")

    # 步骤4：保存数据库
    print('\n步骤4: 保存数据库')
    save_database(db)

    print('\n' + '=' * 60)
    print('完成！')
    print('=' * 60)
    print(f"数据库文件: {Config.DATABASE_FILE}")

if __name__ == '__main__':
    main()

File: test_parse_response.js
/**
 * 测试解析 data/response_body.html
 * 包含完整的去重、数据库保存功能
 */

const fs = require('fs');
const path = require('path');

// 配置
const CONFIG = {
  dataDir: path.join(__dirname, 'data'),
  databaseFile: path.join(__dirname, 'data', 'play_history_db.json')
};

// 确保数据目录存在
function ensureDataDir() {
  if (!fs.existsSync(CONFIG.dataDir)) {
    fs.mkdirSync(CONFIG.dataDir, { recursive: true });
  }
}

// 生成记录唯一ID
function generateRecordId(record) {
  if (record.trackId) {
    return `track_${record.trackId}`;
  }
  // 备用方案：使用曲目+难度+日期+达成率组合
  const key = `${record.title}_${record.diff}_${record.date}_${record.percentage}`;
  return `combo_${Buffer.from(key).toString('base64').replace(/[=+/]/g, '')}`;
}

// 加载数据库
function loadDatabase() {
  ensureDataDir();

  try {
    if (fs.existsSync(CONFIG.databaseFile)) {
      const data = JSON.parse(fs.readFileSync(CONFIG.databaseFile, 'utf8'));
      console.log(`✅ 数据库加载成功，已有 ${data.records.length} 条记录\n`);
      return data;
    }
  } catch (err) {
    console.log(`⚠️  数据库加载失败: ${err.message}，将创建新数据库\n`);
  }

  // 返回空数据库结构
  return {
    version: 1,
    createdAt: new Date().toISOString(),
    lastUpdated: null,
    totalFetches: 0,
    records: [],
    recordIndex: {} // 用于快速查找的索引
  };
}

// 保存数据库
function saveDatabase(db) {
  ensureDataDir();
  db.lastUpdated = new Date().toISOString();
  fs.writeFileSync(CONFIG.databaseFile, JSON.stringify(db, null, 2));
  console.log(`✅ 数据库已保存，共 ${db.records.length} 条记录`);
}

// 将新记录添加到数据库（去重）
function addRecordsToDatabase(db, newRecords) {
  const addedRecords = [];
  let duplicateCount = 0;

  for (const record of newRecords) {
    const recordId = generateRecordId(record);

    // 检查是否已存在
    if (db.recordIndex[recordId]) {
      duplicateCount++;
      continue;
    }

    // 添加元数据
    const enrichedRecord = {
      ...record,
      _id: recordId,
      _fetchedAt: new Date().toISOString()
    };

    // 添加到数据库
    db.records.push(enrichedRecord);
    db.recordIndex[recordId] = true;
    addedRecords.push(enrichedRecord);
  }

  db.totalFetches++;

  return { addedRecords, duplicateCount };
}

// 解析HTML
function parseRecordsHtml(html) {
  console.log(`HTML长度: ${html.length} 字符\n`);

  const records = [];

  // 匹配每个游玩记录块
  const recordBlockRegex = /<div class="p_10 t_l f_0 v_b">([\s\S]*?)(?=<div class="p_10 t_l f_0 v_b">|<div class="f_0">[\s\S]*?<\/footer>)/g;

  let match;
  while ((match = recordBlockRegex.exec(html)) !== null) {
    const block = match[1];

    try {
      // 提取难度
      const diffMatch = block.match(/diff_(\w+)\.png/);
      const diffMap = { 'basic': 'Basic', 'advanced': 'Advanced', 'expert': 'Expert', 'master': 'Master', 'remaster': 'Re:Master' };
      const diff = diffMatch ? (diffMap[diffMatch[1]] || diffMatch[1]) : 'Unknown';

      // 提取日期时间
      const dateMatch = block.match(/<span class="v_b">(\d{4}\/\d{2}\/\d{2} \d{2}:\d{2})<\/span>/);
      const date = dateMatch ? dateMatch[1] : '';

      // 提取曲目名称
      const titleMatch = block.match(/clear\.png"[^>]*\/>([\s\S]*?)<\/div>/);
      let title = '';
      if (titleMatch) {
        title = titleMatch[1].replace(/<[^>]+>/g, '').trim();
        // 解码HTML实体
        title = title.replace(/&#039;/g, "'")
                    .replace(/&quot;/g, '"')
                    .replace(/&amp;/g, '&')
                    .replace(/&lt;/g, '<')
                    .replace(/&gt;/g, '>');
      }

      // 提取达成率
      const achieveMatch = block.match(/playlog_achievement_txt[^>]*>(\d+)<span[^>]*>\.([\d]+)%<\/span>/);
      let percentage = '';
      if (achieveMatch) {
        percentage = `${achieveMatch[1]}.${achieveMatch[2]}`;
      }

      // 提取评级
      const rateMatch = block.match(/playlog\/(\w+)\.png[^>]*class="playlog_scorerank"/);
      let rate = '';
      if (rateMatch) {
        const rateMap = {
          'sssp': 'SSS+', 'sss': 'SSS', 'ssp': 'SS+', 'ss': 'SS',
          'sp': 'S+', 's': 'S', 'aaa': 'AAA', 'aa': 'AA', 'a': 'A',
          'bbb': 'BBB', 'bb': 'BB', 'b': 'B', 'c': 'C', 'd': 'D'
        };
        rate = rateMap[rateMatch[1].toLowerCase()] || rateMatch[1].toUpperCase();
      }

      // 提取DX分数
      const dxMatch = block.match(/white p_r_5 f_15 f_r">([\d,]+)\s*\/\s*([\d,]+)<\/div>/);
      let dxscore = '', dxmax = '';
      if (dxMatch) {
        dxscore = dxMatch[1].replace(/,/g, '');
        dxmax = dxMatch[2].replace(/,/g, '');
      }

      // 提取FC状态
      let fc = null;
      const fcMatch = block.match(/playlog\/(fc_|ap)(\w*)\.png/);
      if (fcMatch && !fcMatch[0].includes('dummy')) {
        const fcType = fcMatch[1] + fcMatch[2];
        const fcMap = { 'ap': 'AP', 'app': 'AP+', 'fc': 'FC', 'fcp': 'FC+' };
        fc = fcMap[fcType.toLowerCase()] || fcType.toUpperCase();
      }

      // 提取FS状态
      let fs = null;
      const fsMatch = block.match(/playlog\/(sync_|fs|fsd)(\w*)\.png/);
      if (fsMatch && !fsMatch[0].includes('dummy')) {
        const fsType = fsMatch[1] + fsMatch[2];
        const fsMap = { 'sync': 'SYNC', 'fs': 'FS', 'fsp': 'FS+', 'fsd': 'FDX', 'fsdp': 'FDX+' };
        fs = fsMap[fsType.toLowerCase()] || fsType.toUpperCase();
      }

      // 提取记录ID
      const idMatch = block.match(/name="idx" value="([^"]+)"/);
      const trackId = idMatch ? idMatch[1] : '';

      // 是否是DX谱面
      const isDx = block.includes('music_dx.png');

      // 是否New Record
      const isNewRecord = block.includes('newrecord.png');

      if (title) {
        records.push({
          title,
          diff,
          percentage,
          rate,
          date,
          fc,
          fs,
          dxscore,
          dxmax,
          trackId,
          isDx,
          isNewRecord
        });
      }
    } catch (e) {
      console.log(`⚠️  解析单条记录失败: ${e.message}`);
    }
  }

  return records;
}

// 主函数
function main() {
  console.log('='.repeat(60));
  console.log('测试解析 data/response_body.html（带去重功能）');
  console.log('='.repeat(60));
  console.log();

  const htmlFile = path.join(__dirname, 'data', 'response_body.html');

  if (!fs.existsSync(htmlFile)) {
    console.error(`❌ 文件不存在: ${htmlFile}`);
    process.exit(1);
  }

  // 步骤1：加载数据库
  console.log('步骤1: 加载历史数据库');
  const db = loadDatabase();

  // 步骤2：读取和解析HTML
  console.log('步骤2: 读取和解析HTML');
  const html = fs.readFileSync(htmlFile, 'utf8');
  const records = parseRecordsHtml(html);

  if (records.length === 0) {
    console.log('⚠️  未找到任何记录');
    process.exit(1);
  }

  console.log(`✅ 成功解析 ${records.length} 条游玩记录\n`);

  // 步骤3：去重并添加到数据库
  console.log('步骤3: 去重并更新数据库');
  const { addedRecords, duplicateCount } = addRecordsToDatabase(db, records);

  console.log(`本次解析: ${records.length} 条`);
  console.log(`新增记录: ${addedRecords.length} 条`);
  console.log(`重复记录: ${duplicateCount} 条（已跳过）`);
  console.log(`数据库总计: ${db.records.length} 条\n`);

  // 显示新增记录
  if (addedRecords.length > 0) {
    console.log('='.repeat(60));
    console.log('新增记录（前10条）');
    console.log('='.repeat(60));
    console.log();

    const displayCount = Math.min(10, addedRecords.length);
    for (let i = 0; i < displayCount; i++) {
      const r = addedRecords[i];
      console.log(`${i + 1}. ${r.title}`);
      console.log(`   难度: ${r.diff} | 日期: ${r.date}`);
      console.log(`   达成率: ${r.percentage}% | 评级: ${r.rate}`);
      console.log(`   DX分数: ${r.dxscore}/${r.dxmax}`);
      console.log();
    }

    if (addedRecords.length > displayCount) {
      console.log(`... 还有 ${addedRecords.length - displayCount} 条新记录\n`);
    }
  } else {
    console.log('ℹ️  所有记录都已存在，没有新增记录\n');
  }

  // 统计信息
  console.log('='.repeat(60));
  console.log('统计信息（数据库总计）');
  console.log('='.repeat(60));

  const stats = {
    total: db.records.length,
    difficulties: {},
    ratings: {},
    fcCount: 0,
    fsCount: 0,
    newRecords: 0,
    dxCount: 0
  };

  for (const r of db.records) {
    stats.difficulties[r.diff] = (stats.difficulties[r.diff] || 0) + 1;
    if (r.rate) stats.ratings[r.rate] = (stats.ratings[r.rate] || 0) + 1;
    if (r.fc) stats.fcCount++;
    if (r.fs) stats.fsCount++;
    if (r.isNewRecord) stats.newRecords++;
    if (r.isDx) stats.dxCount++;
  }

  console.log(`总记录数: ${stats.total}`);
  console.log(`DX谱面: ${stats.dxCount}`);
  console.log(`新纪录: ${stats.newRecords}`);
  console.log(`FC数量: ${stats.fcCount}`);
  console.log(`FS数量: ${stats.fsCount}`);

  console.log('\n难度分布:');
  const diffOrder = ['Basic', 'Advanced', 'Expert', 'Master', 'Re:Master'];
  for (const diff of diffOrder) {
    if (stats.difficulties[diff]) {
      console.log(`  ${diff}: ${stats.difficulties[diff]}`);
    }
  }

  console.log('\n评级分布:');
  const ratingOrder = ['SSS+', 'SSS', 'SS+', 'SS', 'S+', 'S', 'AAA', 'AA', 'A'];
  for (const rate of ratingOrder) {
    if (stats.ratings[rate]) {
      console.log(`  ${rate}: ${stats.ratings[rate]}`);
    }
  }

  // 步骤4：保存数据库
  console.log('\n步骤4: 保存数据库');
  saveDatabase(db);

  console.log('\n='.repeat(60));
  console.log('完成！');
  console.log('='.repeat(60));
  console.log(`数据库文件: ${CONFIG.databaseFile}`);
}

main();
